src/lib/fields_derivers/fields_derivers.ml

```diff
@@ -0,0 +1,188 @@
+module type Deriver_basic_intf = sig
+  module Input : sig
+    type 'input_type t
+  end
+
+  module Creator : sig
+    type 'input_type t
+  end
+
+  module Output : sig
+    type 'output_type t
+  end
+
+  module Accumulator : sig
+    type 'input_type t
+  end
+
+  val init : unit -> 'input_type Accumulator.t
+
+  val add_field :
+       'f Input.t
+    -> ( [< `Read | `Set_and_create ]
+       , 'input_type
+       , 'f )
+       Fieldslib.Field.t_with_perm
+    -> 'input_type Accumulator.t
+    -> ('input_type Creator.t -> 'f) * 'input_type Accumulator.t
+
+  val finish :
+       ('input_type Creator.t -> 'input_type) * 'input_type Accumulator.t
+    -> 'input_type Output.t
+
+  val int_ : int Input.t
+
+  val string_ : string Input.t
+
+  val bool_ : bool Input.t
+
+  val list_ : 'a Input.t -> 'a list Input.t
+end
+
+module type Deriver_intf = sig
+  include Deriver_basic_intf
+
+  module Step : sig
+    type ('a, 'input_type, 'f) t =
+         ( ([< `Read | `Set_and_create ] as 'a)
+         , 'input_type
+         , 'f )
+         Fieldslib.Field.t_with_perm
+      -> 'input_type Accumulator.t
+      -> ('input_type Creator.t -> 'f) * 'input_type Accumulator.t
+  end
+
+  module Prim : sig
+    val int : ('a, 'input_type, int) Step.t
+
+    val string : ('a, 'input_type, string) Step.t
+
+    val bool : ('a, 'input_type, bool) Step.t
+
+    val list : 'l Input.t -> ('a, 'input_type, 'l list) Step.t
+  end
+end
+
+module Make (D : Deriver_basic_intf) :
+  Deriver_intf
+    with module Accumulator = D.Accumulator
+     and module Input = D.Input
+     and module Creator = D.Creator
+     and module Output = D.Output = struct
+  include D
+
+  module Step = struct
+    type ('a, 'input_type, 'f) t =
+         ( ([< `Read | `Set_and_create ] as 'a)
+         , 'input_type
+         , 'f )
+         Fieldslib.Field.t_with_perm
+      -> 'input_type Accumulator.t
+      -> ('input_type Creator.t -> 'f) * 'input_type Accumulator.t
+  end
+
+  module Prim = struct
+    let int fd acc = add_field D.int_ fd acc
+
+    let string fd acc = add_field D.string_ fd acc
+
+    let bool fd acc = add_field D.bool_ fd acc
+
+    let list l fd acc = add_field (D.list_ l) fd acc
+  end
+end
+
+let under_to_camel s =
+  let open Core_kernel in
+  let ws = String.split s ~on:'_' in
+  match ws with
+  | [] ->
+      ""
+  | w :: ws ->
+      w :: (ws |> List.map ~f:String.capitalize) |> String.concat ?sep:None
+
+let%test_unit "under_to_camel works as expected" =
+  let open Core_kernel in
+  [%test_eq: string] "fooHello" (under_to_camel "foo_hello") ;
+  [%test_eq: string] "fooHello" (under_to_camel "foo_hello___")
```

Comment:
you might want a test case with a leading underscore, which could occur in practice, if the name is ignored